benchmarks/libseff/resume_nontail/main.c

```diff
@@ -0,0 +1,67 @@
+#include "seff.h"
+#include <math.h>
+#include <stdio.h>
+
+DEFINE_EFFECT(op, 0, void, { int64_t x; });
+
+static inline int64_t binaryOperation(int64_t x, int64_t y) { 
+  return (labs(x - (503 * y) + 37)) % 1009;
+}
+
+static int64_t handleOpRec(seff_coroutine_t *k) {
+  seff_request_t req = seff_handle(k, NULL, HANDLES(op));
+  switch (req.effect) {
+    CASE_EFFECT(req, op, {
+      return binaryOperation(payload.x, handleOpRec(k));
+    })
+    CASE_RETURN(req, {
+      return (int64_t) payload.result;
+    })
+  };
+  return -1;
+}
+
+typedef struct loop_args_t {
+  int64_t n;
+  int64_t s;
+} loop_args_t;
+
+// Note: using a for loop instead of recursion because
+// it is the idiomatic way of doing loops in C
+static void* loop(void* parameter) {
+  loop_args_t* args = (loop_args_t*) parameter;
+  for (int i = args->n; i > 0; i--) {
+    PERFORM(op, i);
+  }
+  return (void*) args->s;
+}
```

Comment:
Similar to `product_early`, shouldn't this function be recursive?

Reply:
I think this one is a bit different. In the description, the goal of the benchmark is to have a deep resumption stack, so the handler should call the continuation in a non-tail recursive way. To compare, `product_early` requires multiple stack frames, so the coroutine needs to be non-tail recursive.

However, consistency is good, so I tried to rewrite "resume_nontail" in a tail-recursive function. However, it seems that clang, even with optimizations on, cannot optimize away the function call. Since the other languages probably optimize the tail-recursive loop into a simple loop, I wonder if keeping the C code as a loop would not be more fair.

Here is my code, let me know if you can figure out why it is not optimizing it.
![image](https://github.com/effect-handlers/effect-handlers-bench/assets/35728891/bf8e5246-5587-4d1a-b40b-ffe3b304b6d8)

Reply:
You're right, I misread the description, I'd keep the loop then.

Interesting bit, I've been playing a bit in godbolt (https://godbolt.org/z/65jcTab9G):
* tail call seems to be not automatically optimised due to `PERFORM(op, args->n)` escaping a local variable, `payload`, which is generated by the macro
* Adding a `__attribute__((musttail))` forces the TCO and it works fine (performance is fine as well)

However, point 1 is very impoortant, since it means that `PERFORM(...)` disables some optimisations, we should look at this in more detail, thank you for the finding!

Reply:
That is really interesting! Is there a good reason why `PERFORM` declares its own `__payload` variable? I was just wondering if the two lines in `PERFORM` could not be combined somehow. In the godbolt example, `g(a)` seems to be optimized by the compiler.

Reply:
Yes, `g(a)` gets optimised because it doesn't escape any local addresses (it would be different if it were `g(&a)`).

`PERFORM` declares its own `__payload` just to make it easier for the user, but that's not the main issue, I'd encourage you to manually expand that macro and try to (still using `seff_perform`) get back the TCO by moving things around.

The question is, where does the payload of the effect should be stored, there are a few choices:
* in the stack frame of the performing function, this is what we currently do, which seems to disbale TCO, as you noticed
* in the heap, but we really don't want to do this since it'd probably be very slow
* somehow at the stack frame of the handler, that is, before calling `seff_handle` some space should be allocated for the payload of each hadnled effect, and registered somewhere. The problem with this is that, in order for it to work, the performing operation should get an address for this space (dynamically) and copy the payload in there, which would of course mean unnecessary indirections. 
* Maybe some other way, but it's hard since the payloads of the effects have arbitrary sizes.

Reply:
I will reply to you privately and resolve this conversation, only because it has grown into a discussion related to libseff specifically rather than the benchmarks :+1: